svm/src/account_loader.rs

```diff
@@ -95,6 +95,264 @@ pub struct FeesOnlyTransaction {
+        rollback_accounts: &RollbackAccounts,
+    ) {
+        let fee_payer_address = message.fee_payer();
+        match rollback_accounts {
+            RollbackAccounts::FeePayerOnly { fee_payer_account } => {
+                self.update_account(fee_payer_address, fee_payer_account);
+            }
+            RollbackAccounts::SameNonceAndFeePayer { nonce } => {
+                self.update_account(nonce.address(), nonce.account());
+            }
+            RollbackAccounts::SeparateNonceAndFeePayer {
+                nonce,
+                fee_payer_account,
+            } => {
+                self.update_account(nonce.address(), nonce.account());
+                self.update_account(fee_payer_address, fee_payer_account);
+            }
+        }
+    }
+
+    fn update_accounts_for_successful_tx(
+        &mut self,
+        message: &impl SVMMessage,
+        transaction_accounts: &[TransactionAccount],
+    ) {
+        for (i, (address, account)) in (0..message.account_keys().len()).zip(transaction_accounts) {
+            if !message.is_writable(i) {
+                continue;
+            }
+
+            // Accounts that are invoked and also not passed as an instruction
+            // account to a program don't need to be stored because it's assumed
+            // to be impossible for a committable transaction to modify an
+            // invoked account if said account isn't passed to some program.
+            if message.is_invoked(i) && !message.is_instruction_account(i) {
+                continue;
+            }
+
+            self.update_account(address, account);
+        }
+    }
+
+    fn update_account(&mut self, account_key: &Pubkey, account: &AccountSharedData) {
+        // If an account has been dropped, we must hide the data from any future transactions.
+        // We NEVER remove items from cache, or else we will fetch stale data from accounts-db.
+        // Because `update_account()` is only called on writable accounts using the same rules
+        // as commit, this (correctly) does not shadow rent-paying zero-lamport read-only accounts.
+        let account = if account.lamports() == 0 {
+            &AccountSharedData::default()
+        } else {
+            account
+        };
+
+        // Accounts are inspected during loading, and we only update writable accounts.
+        // Ergo, `inspected_as_writable` is always true.
+        self.account_cache
+            .entry(*account_key)
+            .and_modify(|cache_item| {
+                debug_assert!(cache_item.inspected_as_writable);
+                cache_item.account = account.clone();
+            })
+            .or_insert_with(|| AccountCacheItem {
+                account: account.clone(),
+                inspected_as_writable: true,
+            });
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub(crate) enum AccountUsagePattern {
+    Writable,
+    Instruction,
+    WritableInstruction,
+    ImmutableInvisible,
+}
```

Comment:
im not married to the name `ImmutableInvisible` but it seemed most appropriate. i couldnt call it `Program` because it is possible (albeit useless) for people to manually add accounts to a message that no instructions use. i _believe_ extra accounts can also be added as writable, but there should be no edge cases where the normal writable flow is inapproproate